stats: Improved sequencing documentation for server-side stats events and added tests.

RyanBlaney opened the pull request:
#### Response to Issue #7824 

RELEASE NOTES:

- **Documentation Clarification**: Clarified the expected sequencing of stats events in the server-side stats.Handler.
- **New Test Cases**: Added tests to verify the event sequencing for various RPC types, ensuring consistency and accuracy in server-side stats handling.

RyanBlaney commented on stats/stats_test.go:
```diff
@@ -119,12 +150,25 @@ type testServer struct {
 }
 
-func (s *testServer) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*testpb.SimpleResponse, error) {
+func (s *testServer) UnaryCall(
+	ctx context.Context,
+	in *testpb.SimpleRequest,
+) (*testpb.SimpleResponse, error) {
```
Got it! I'll disable my autoformatter going forward or reconfigure it to follow the Google style guide.

RyanBlaney commented on stats/stats_test.go:
```diff
@@ -786,8 +884,14 @@ func checkConnEnd(t *testing.T, d *gotData) {
 
+type event struct {
+	eventType string
+	timestamp time.Time
```
The timestamp was included for debugging purposes in case of test failures, since it could help identify delays or other issues. If you feel this isn’t valuable, I can to remove it for simplicity.

RyanBlaney commented:
I have fixed the wrapping errors for formatting and I removed the timestamps from the event structs. Thank you for your feedback @arjan-bal, is there anything else you need me to do?

purnesh42H commented on stats/stats.go:
```diff
@@ -36,7 +36,49 @@ type RPCStats interface {
 }
 
+// InHeader contain stats when the header is received.
+//  First event in the server side event sequence.
```
nix the extra space. Applicable everywhere. If you want to move the sentence to new line, you need to put a blank line in between. If you want bullets, you can use markdown style "-"

purnesh42H commented on stats/stats.go:
```diff
@@ -98,31 +140,6 @@ func (s *InPayload) IsClient() bool { return s.Client }
 func (s *InPayload) isRPCStats() {}
 
-// InHeader contains stats when a header is received.
-type InHeader struct {
```
@RyanBlaney we should just edit the documentation in current struct instead of deleting it entirely and re-creating. Also, restrict the documentation to struct instead of writing the whole sequence. Mentioning that it is the first event  on server side after rpc is received is fine.

dfawley replied:
> Also, restrict the documentation to struct instead of writing the whole sequence

I think it would be nice to show the whole sequence somewhere in our documentation.  That would probably belong as a package-level comment.  But not mandatory for this PR if you don't want to do it.  Note that the possible sequences can't simply be a list "A then B then C" because the ordering is more complex (e.g. the server headers could be received before/during/after client out payloads).

purnesh42H commented on stats/stats_test.go:
```diff
@@ -136,7 +197,14 @@ func (s *testServer) UnaryCall(ctx context.Context, in *testpb.SimpleRequest) (*
 func (s *testServer) FullDuplexCall(stream testgrpc.TestService_FullDuplexCallServer) error {
 	if err := stream.SendHeader(testHeaderMetadata); err != nil {
-		return status.Errorf(status.Code(err), "%v.SendHeader(%v) = %v, want %v", stream, testHeaderMetadata, err, nil)
+		return status.Errorf(
```
golang doesn't restrict columns for code. Having it in one is fine.

purnesh42H commented on stats/stats_test.go:
```diff
@@ -81,6 +83,55 @@ var (
 	// The id for which the service handler should return error.
 	errorID int32 = 32202
+
+    // Server Stats
```
Please fix the documentation everywhere. Remove extra spaces at start and end etc.

RyanBlaney replied:
Everything should be resolved now. Thank you for your help!

RyanBlaney commented:
I added a note for the StartServer function in the stats_test.go. It uses a few deprecated opts, so if you want me to resolve these issues I can do that. Also there is a TODO that got left for testing the OutHeader. If you need me to test that I can do that as well.

purnesh42H commented on stats/stats.go:
```diff
@@ -37,6 +37,9 @@ type RPCStats interface {
 }
 
 // Begin contains stats when an RPC attempt begins.
+// This event is called AFTER the InHeader event, as headers must
```
why is AFTER caps?

Also, you don't have to begin from next line unless you reaching the max columns 80 or if you want to have a new line then you have to put a blank line in between two sentences.

purnesh42H replied:
nit: `InHeader`

RyanBlaney replied:
Got it! Cleaning this up in the next commit.

purnesh42H commented on stats/stats.go:
```diff
@@ -98,7 +101,10 @@ func (s *InPayload) IsClient() bool { return s.Client }
+// InHeader contain stats when the header is received.
+// 
+// First event in the server side event sequence.
+// Follows last OutPayload for server side events.
```
Same thing. Can start in same line. `It is the first event at the server after receiving the RPC. It is followed by `OutPayload` server event.`

purnesh42H commented on stats/stats.go:
```diff
@@ -123,6 +129,7 @@ func (s *InHeader) IsClient() bool { return s.Client }
 
 func (s *InHeader) isRPCStats() {}
 
+
```
nix new line

RyanBlaney replied:
What does nix mean in this context? I'm assuming you mean delete.

purnesh42H commented on stats/stats_test.go:
```diff
@@ -59,8 +59,10 @@ func init() {
 
-type connCtxKey struct{}
-type rpcCtxKey struct{}
+type (
```
We don't declare types like that anywhere in our codebase so let's revert this as its not relevant to the change?

RyanBlaney replied:
My bad. Remnant of an gofmt mistake.

purnesh42H commented on stats/stats_test.go:
```diff
@@ -81,6 +83,52 @@ var (
 	// The id for which the service handler should return error.
 	errorID int32 = 32202
+    // Server Stats
+	// Ensure that Unary RPC server stats events are logged in the correct order.
```
variable name is self explanatory. If at all we want to put comment here, something like this is more suitable, `To verify if the Unary RPC server stats events are logged in the correct order.`

purnesh42H commented on stats/stats_test.go:
```diff
@@ -81,6 +83,52 @@ var (
 	}
 	// The id for which the service handler should return error.
 	errorID int32 = 32202
+    // Server Stats
```
why do we need this extra comment `ServerStats`? Probably just remove since you have a sentence already for each. Applicable everywhere

purnesh42H commented on stats/stats_test.go:
```diff
@@ -326,7 +376,11 @@ func (te *test) doUnaryCall(c *rpcConfig) (*testpb.SimpleRequest, *testpb.Simple
 
 	tCtx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	resp, err = tc.UnaryCall(metadata.NewOutgoingContext(tCtx, testMetadata), req, grpc.WaitForReady(!c.failfast))
```
golang doesn't restrict columns for code so let's avoid the changes that are not relevant to the issue. Applicable everywhere.

RyanBlaney replied:
Remnant of an autoformatting mistake.

purnesh42H commented:
@RyanBlaney could you please revert the refactor/cleanup changes that are unrelated to the issue? Its hard to review what is changing.

RyanBlaney commented:
I fixed the autoformatting mistakes and cleaned up the comments. I will not use an autoformatter going forward, since it caused me a headache. Is there anything else needed in terms of testing?

purnesh42H commented:
> I fixed the autoformatting mistakes and cleaned up the comments. I will not use an autoformatter going forward, since it caused me a headache. Is there anything else needed in terms of testing?

Thanks, I will take a look. Looks like you have some vet check failures. Could you run the vet check locally and fix https://github.com/grpc/grpc-go/blob/master/CONTRIBUTING.md ?

RyanBlaney commented:
The vet errors should be fixed now. Happy holidays!

RyanBlaney commented:
I'm still failing one test for some reason. I didn't see that locally. I'll look into it when I get the chance.

purnesh42H commented on stats/stats.go:
```diff
@@ -36,7 +36,10 @@ type RPCStats interface {
 
-// Begin contains stats when an RPC attempt begins.
+// Begin contains stats when an RPC attempt begins. This event is called after
+// the InHeader event, as headers must be processed before the RPC lifecycle
```
nit `InHeader`

RyanBlaney replied:
What specifically is the issue here? Is it related to style?

purnesh42H replied:
Names from code should preferably be within backticks

purnesh42H commented on stats/stats.go:
```diff
@@ -98,7 +101,9 @@ func (s *InPayload) IsClient() bool { return s.Client }
 
-// InHeader contains stats when a header is received.
+// InHeader contain stats when the header is received. It is the first event in
+// the server after receiving the RPC. It is followed by the OutPayload
```
`OutPayload`

RyanBlaney replied:
Can you clarify what you are asking for here?

purnesh42H replied:
Names from code should preferably be within backticks

purnesh42H replied:
Also, isn't it followed by Begin?

purnesh42H commented on stats/stats_test.go:
```diff
@@ -81,6 +81,50 @@ var (
 	errorID int32 = 32202
+	// To verify if the Unary RPC server stats events are logged in the
+	// correct order.
+	expectedUnarySequence = []string{
```
this can be initialized within the test itself because we are not using it anywhere else. Also, convention is to use `want/wanted` instead of `expected`

RyanBlaney replied:
Got it! I will keep this in mind going forward. Thank you!

purnesh42H commented on stats/stats_test.go:
```diff
@@ -81,6 +81,50 @@ var (
+	}
+	// To verify if the Client Stream RPC server stats events are logged in the
+	// correct order.
+	expectedClientStreamSequence = []string{
```
same. can be moved to the respective test. Also, `wantClientStreamSequence `

purnesh42H commented on stats/stats_test.go:
```diff
@@ -81,6 +81,50 @@ var (
+	}
+	// To verify if the Server Stream RPC server stats events are logged in the
+	// correct order.
+	expectedServerStreamSequence = []string{
```
same. can be moved to respective test. s/expectedServerStreamSequence/wantServerStreamSequence

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1598,77 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	verifyEventSequence(t, h.events, expectedServerStreamSequence)
```
can't we use cmp.Equal? Here we do care about order being same as well.

RyanBlaney replied:
I replaced with cmp.Equal although I used it in the helper function for reusability.

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1598,77 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	verifyEventSequence(t, h.events, expectedClientStreamSequence)
```
same. Prefer cmp.Equal

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1598,77 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	// Verify sequence
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	verifyEventSequence(t, h.events, expectedUnarySequence)
```
same. cmp.Equal

purnesh42H commented on stats/stats_test.go:
```diff
@@ -800,13 +807,41 @@ func (h *statshandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) conte
+	switch s.(type) {
+	case *stats.Begin:
+		h.recordEvent("Begin")
+	case *stats.InHeader:
```
may be InHeader should be first case to be consistent with the order of occurence

purnesh42H replied:
@RyanBlaney ping on this

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1554,122 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+		gotEventTypes[i] = e.eventType
+	}
+	if !cmp.Equal(gotEventTypes, expected) {
+		t.Errorf("Event sequence mismatch (-got +expected):\n%s", cmp.Diff(gotEventTypes, expected))
```
s/expected/want

purnesh42H replied:
`Diff (-got +expected)` since you are showing the diff

RyanBlaney replied:
I'm assuming by added the vim command you want me to replace all occurrences of 'expected.'

RyanBlaney replied:
It is followed by Begin for server-side stats events. It follows the OutPayload for client-side events. I need to clarify this.

purnesh42H replied:
No I meant to ask to edit the error message to show "Diff (-got +expected)" :)

RyanBlaney replied:
Ohhhh! I see. My bad. Thanks for clarifying.

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1554,122 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+
+// verifyEventSequence verifies that a sequence of recorded events matches
+// the expected sequence.
+func verifyEventSequence(t *testing.T, got []event, expected []string) {
```
s/expected/want

RyanBlaney replied:
There are lines that I didn't write that contain the word 'expected.' Should those be changed too?

purnesh42H commented on stats/stats_test.go:
```diff
@@ -786,8 +788,13 @@ func checkConnEnd(t *testing.T, d *gotData) {
+
 type statshandler struct {
 	mu      sync.Mutex
+	events  []event
```
it looks like we just need list of event types? Can't we directly use []string?

RyanBlaney replied:
Yes that should suffice.

purnesh42H reviewed (commented):
Looking close to LGTM. Few last set of comments.

arjan-bal reviewed (commented):
Unblocking for other reviewers to approve.

purnesh42H commented on stats/stats.go:
```diff
@@ -36,7 +36,11 @@ type RPCStats interface {
 }
 
-// Begin contains stats when an RPC attempt begins.
+// Begin contains stats recorded when an RPC attempt begins. This event is
```
How about something like this? Removing some repititions.

// Begin marks the start of an RPC attempt for stats recording.
//
// - Server-side: Triggered after `InHeader`, as headers are processed before the RPC lifecycle begins.
// - Client-side: The first stats event recorded.

RyanBlaney replied:
Yeah that looks good! The bullets seem easier on the eyes in my editor.

purnesh42H commented on stats/stats.go:
```diff
@@ -98,7 +102,9 @@ func (s *InPayload) IsClient() bool { return s.Client }
 func (s *InPayload) isRPCStats() {}
 
-// InHeader contains stats when a header is received.
+// InHeader contains stats recorded when the header is received. It is the
```
same here

// InHeader records stats upon header reception.
//
// - Server-side: The first stats event after the RPC request is received.
// - Client-side: Occurs after the `OutPayload` event (after the request payload is sent).

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	defer h.mu.Unlock()
+	// To verify if the Unary RPC server stats events are logged in the
+	// correct order.
+	wantedUnarySequence := []string{
```
nit: wantUnarySequence

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	}
+	// Allow time for events to propagate
+	time.Sleep(50 * time.Millisecond)
+	// Verify sequence
```
nit: new line after time.Sleep

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
```
nix new line

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
```
nix new line

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+
+	// To verify if the Server Stream RPC server stats events are logged in the
+	// correct order.
+	wantedServerStreamSequence := []string{
```
nit: wantServerStreamSequence

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	defer h.mu.Unlock()
+	// To verify if the Client Stream RPC server stats events are logged in the
+	// correct order.
+	wantedClientStreamSequence := []string{
```
nit: wantClientStreamSequence

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	copy(gotEventTypes, got)
+
+	if !cmp.Equal(gotEventTypes, want) {
+		t.Errorf("Event sequence mismatch (-got +want):\n%s", cmp.Diff(gotEventTypes, want))
```
"Event sequence mismatch. Diff (-got +want):\n%s"

purnesh42H commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,121 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+	t.Helper()
+	// Extract event types from `got` for comparison.
+	gotEventTypes := make([]string, len(got))
+	copy(gotEventTypes, got)
```
do we need to copy?

purnesh42H reviewed (commented):
Some more non blocking comments. Adding @dfawley for second review

purnesh42H reviewed (approved):
lgtm to unblock

dfawley commented on stats/stats_test.go:
```diff
@@ -433,7 +435,7 @@ func (te *test) doServerStreamCall(c *rpcConfig) (*testpb.StreamingOutputCallReq
 	}
 }
 
-type expectedData struct {
```
This is just code churn and an arbitrary style decision. Can you please revert it?

RyanBlaney replied:
Got it! Do you recommend I revert other instances of 'want' back to 'expected'? Specifically the ones that I introduced in my tests.

dfawley replied:
I do prefer consistency, at least within a file, but I'm not all that bothered about want vs. expected.

dfawley commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,117 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	// Allow time for events to propagate
+	time.Sleep(50 * time.Millisecond)
```
This is a flaky test waiting to happen.

Please see how we do this kind of thing in other places.  Basically:

- Loop until the test's ~10s deadline is reached
- check the condition you are trying to verify
- If it wasn't found, then sleep a short time (e.g. 10ms) and repeat

RyanBlaney replied:
Thanks for the feedback! I'll be more careful in future tests.

RyanBlaney commented on stats/stats.go:
```diff
@@ -98,7 +102,10 @@ func (s *InPayload) IsClient() bool { return s.Client }
 func (s *InPayload) isRPCStats() {}
 
-// InHeader contains stats when a header is received.
+// InHeader records stats upon header reception.
```
The language seemed off when I originally read it, so it definitely needed a revision. I'll revert 'records' to 'contains', then I'll match the changes on the other structs.

RyanBlaney replied:
I made some revisions to the phrasing for consistency. Let me know what you think. I'm still working on updating the test logic.

purnesh42H commented:
@RyanBlaney is it ready for re-review? I see that you have commits for the last comments. Also, since its been called out, you can revert the style changes for want/expect in existing code but keep want/wantData in the new code that you have written

RyanBlaney commented:
Yes, it should be ready! Everything new that I added is using want instead of expected.

dfawley commented on stats/stats_test.go:
```diff
@@ -800,13 +803,41 @@ func (h *statshandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) conte
 func (h *statshandler) HandleRPC(ctx context.Context, s stats.RPCStats) {
+	switch s.(type) {
+	case *stats.Begin:
+		h.recordEvent("Begin")
```
Could you define constants for these strings please?

dfawley commented on stats/stats_test.go:
```diff
@@ -1519,3 +1550,174 @@ func (s) TestStatsHandlerCallsServerIsRegisteredMethod(t *testing.T) {
+
+// TestServerStatsUnaryRPCEventSequence tests that the sequence of server-side stats
+// events for a Unary RPC matches the expected flow.
+func (s) TestServerStatsUnaryRPCEventSequence(t *testing.T) {
```
How are these tests different from what we already have?

https://github.com/grpc/grpc-go/blob/79b6830e4bbe341cbb855c86b1caa5482d9d040a/stats/stats_test.go#L929

RyanBlaney replied:
I was confused since in the original issue, I was informed that there were no tests which test the sequence of stats events, although the existing tests seem to test the sequence. Are we good to just remove them entirely, or is there a specific behavior we are looking to test?

dfawley replied:
Sorry, this was probably a misunderstanding by the person that created the issue.  I believe I asked them to determine/verify if we were testing the ordering, and they may have just assumed that we weren't, or assumed that whoever made the change would do the verifying.

I think we can just remove the new tests, then?  Unless you can identify any gaps, in which case, ideally we'd just extend them.

purnesh42H replied:
@RyanBlaney in case you missed this

RyanBlaney replied:
I saw it, I just had a busy week so I wasn't able to get to this yet. I will just remove them since they aren't testing anything new. I will make a new PR if I decide to extend the current tests. Thank you again for reviewing—both of you!

RyanBlaney commented:
This PR should be complete, although there are a few more changes I want to make for a second PR. 

## Implement all TODO tests

- There are several TODO statements left throughout the code, so I will implement the tests as needed.
- Test `OutTrailer`
- Remove `IsClient` call (2 instances)

## Possible logic error

- The docs state that the CompressedLength should be equal to the Length if compression is disabled.
- The Length in this case is the uncompressed InPayload data

```go
	// TODO: check values of WireLength and RecvTime.
	if st.Length > 0 && st.CompressedLength == 0 {
		t.Fatalf("st.WireLength = %v with non-empty data, want <non-zero>",
			st.CompressedLength)
	}
```

## Fix deprecated tests

```
Diagnostics:
grpc.RPCCompressor is deprecated: use encoding.RegisterCompressor instead. Will be supported throughout 1.x. [default]

Diagnostics:
grpc.NewGZIPCompressor is deprecated: use package encoding/gzip. [default]

Diagnostics:
grpc.RPCDecompressor is deprecated: use encoding.RegisterCompressor instead. Will be supported throughout 1.x. [default]

Diagnostics:
grpc.NewGZIPDecompressor is deprecated: use package encoding/gzip. [default]

Diagnostics:
grpc.WithBlock is deprecated: this DialOption is not supported by NewClient. Will be supported throughout 1.x. [default]

Diagnostics:
grpc.Dial is deprecated: use NewClient instead.  Will be supported throughout 1.x. [default]
```

dfawley commented on stats/stats.go:
```diff
@@ -98,7 +102,10 @@ func (s *InPayload) IsClient() bool { return s.Client }
+// InHeader contains stats about header reception.
+//
+// - Server-side: The first stats event after the RPC request is received.
+// - Client-side: Occurs after the `OutPayload` event (after the request payload is sent).
```
This isn't quite true.  InHeader comes any time after OutHeader.  This is especially true for bidi- or client- streaming RPCs, since OutPayload isn't guaranteed to come right after OutHeader, like it is for unary or server-streaming RPCs.

Maybe exclude the note about client-side here, but you can add a note to `OutHeader` for `Client-side` that says _something like_ `Only occurs after 'Begin', as headers are always the first thing sent on a stream.`

dfawley commented on stats/stats_test.go:
```diff
@@ -800,13 +803,52 @@ func (h *statshandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) conte
 }
 
+// recordEvent records an event in the statshandler along with a timestamp.
+func (h *statshandler) recordEvent(eventType string) {
```
Do we need any of this now?  I don't think anyone reads this field since the new tests were removed.

RyanBlaney replied:
Nope! Forgot to remove this. Thanks!

dfawley commented on stats/stats_test.go:
Could we just revert this file now?  It mostly looks like churn which makes the blame history harder to navigate.

RyanBlaney replied:
Yeah fair enough. Since I removed the tests anyway.

dfawley reviewed (approved):
Thank you!

dfawley commented:
Thanks!  Over to @purnesh42H for a quick second pass.

purnesh42H commented on stats/stats.go:
```diff
@@ -36,7 +36,11 @@ type RPCStats interface {
-// Begin contains stats when an RPC attempt begins.
+// Begin contains stats for the start of an RPC attempt.
+//
+// - Server-side: Triggered after `InHeader`, as headers are processed before the RPC lifecycle begins.
```
nit: restrict comments to 80 cols per line

RyanBlaney replied:
Got it! Thanks!

purnesh42H reviewed (commented):
@RyanBlaney 1 more nit else looks good.
